relayer/chains/cosmos/tx.go

```diff
@@ -649,22 +669,260 @@ func (cc *CosmosProvider) buildMessages(
 	// Build the transaction builder
 	txb, err := txf.BuildUnsignedTx(cMsgs...)
 	if err != nil {
-		return nil, 0, sdk.Coins{}, err
+		return nil, nil, err
 	}
+	return &txf, txb, nil
+}
 
-	if err = tx.Sign(txf, txSignerKey, txb, false); err != nil {
-		return nil, 0, sdk.Coins{}, err
+func convertAddress(addrString string) (*common.Address, error) {
+	cfg := sdk.GetConfig()
+	var addr []byte
+	// try hex, then bech32
+	switch {
+	case common.IsHexAddress(addrString):
+		addr = common.HexToAddress(addrString).Bytes()
+	case strings.HasPrefix(addrString, cfg.GetBech32ValidatorAddrPrefix()):
+		addr, _ = sdk.ValAddressFromBech32(addrString)
+	case strings.HasPrefix(addrString, cfg.GetBech32AccountAddrPrefix()):
+		addr, _ = sdk.AccAddressFromBech32(addrString)
+	default:
+		return nil, fmt.Errorf("expected a valid hex or bech32 address (acc prefix %s), got '%s'", cfg.GetBech32AccountAddrPrefix(), addrString)
 	}
+	to := common.BytesToAddress(addr)
+	return &to, nil
+}
 
-	tx := txb.GetTx()
-	fees = tx.GetFee()
+func convertCoins(prices string) (sdk.Coins, error) {
+	coins, err := sdk.ParseCoinsNormalized(prices)
+	if err != nil {
+		return nil, errorsmod.Wrapf(err, "failed to parse coins")
+	}
+	return coins, nil
+}
 
+func (cc *CosmosProvider) signTxBytes(
+	ctx context.Context,
+	tx authsigning.Tx,
+) ([]byte, error) {
+	var txBytes []byte
 	// Generate the transaction bytes
-	txBytes, err = cc.Cdc.TxConfig.TxEncoder()(tx)
+	if err := retry.Do(func() error {
+		var err error
+		txBytes, err = cc.Cdc.TxConfig.TxEncoder()(tx)
+		if err != nil {
+			return err
+		}
+		return nil
+	}, retry.Context(ctx), rtyAtt, rtyDel, rtyErr); err != nil {
+		return nil, err
+	}
+	return txBytes, nil
+}
+
+func (cc *CosmosProvider) buildMessages(
+	ctx context.Context,
+	txf *tx.Factory,
+	txb client.TxBuilder,
+	txSignerKey string,
+) ([]byte, uint64, sdk.Coins, error) {
+	if err := tx.Sign(*txf, txSignerKey, txb, false); err != nil {
+		return nil, 0, sdk.Coins{}, err
+	}
+	tx := txb.GetTx()
+	txBytes, err := cc.signTxBytes(ctx, tx)
```

Comment:
it seems signed twice here?

Reply:
yes, only call once for buildMessages or buildEvmMessages